### DISCUSSION TITLE
sys/linux IPC_Flags_Bits and Sem_Buf

### DISCUSSION THREAD

**Discussion by AxelArielSaravia** [odin-lang/Odin#3975] [Q&A] [answered]
Hi, I am following the Beej's [Guide to Unix Interproces Communication](https://beej.us/guide/bgipc/) and rewriting the examples in Odin (version dev-2024-07) using the sys/linux package. I encountered some issues with `IPC_Flags` in general because the `IPC_Flags_Bits` enum does not have any enumerable for the last significant 9 bits (well, it has for the 9' and 7' bit) that set the permissions for functions like [msgget()](https://www.man7.org/linux/man-pages/man2/msgget.2.html) or [semget()](https://www.man7.org/linux/man-pages/man2/semget.2.html). **I do not know if this is an error or if there is an Odin way that I am not aware of**. Maybe combining the `IPC_Flags_Bits` with the `IPC_Mode_Bits`??
My solutions are to use `linux.syscall` directly or use `transmute` operator: 
```odin
//IPC_Flags_Bits.IPC_CREAT | 0o666
ret := linux.syscall(linux.SYS_msgget, key, i16(0o1666))
//or
msgid, lerrno := linux.msgget(key, transmute(linux.IPC_Flags)(i16(0o1666)))
```

A similar issue occurs with the Sem_Buf structure. Where the man page for[semop](https://www.man7.org/linux/man-pages/man2/semop.2.html) indicates that the `op` struct member can be positive, negative, or zero, but in the sys/linux package, it is an `IPC_Cmd` enum (🤷🏻‍♂️). **Is the C way obscuring the Odin way, is there something I am missing?**
My solutions are to use `linux.syscall` directly or use `transmute` operator:
```odin
Sembuf :: struct {
    num :u16,
    op  :i16,
    flg :IPC_Flags
}
...
sb := Sembuf{op = 1}
...
lerrno = linux.semop(key, {transmute(linux.Sem_Buf) sb})
```
Thanks

**Comment by flysand7** [comment, marked as answer]
Odin should match the linux syscall interface relatively closely, aside from the following exceptions:

1. Flags are bitsets
2. If the interface assumes concatenation of an integer and a bitset, those are separated into different arguments
3. Pointer + length pairs are replaced with slices in procedure paramters
4. Errno (if present) and the numberic return value are returned as separate arguments.

If you believe that Odin's functions are wrong, feel free to make an issue, I'll fix it within a week (when I have the time). Some of these functions in linux syscall interface weren't tested so I'm not sure if any of them still contain mistakes.

Also please note that the book you have linked uses POSIX interface (a C library), not linux syscalls directly. As of right now, Odin does not provide significant bindings for posix interfaces, and right now we're only implementing the syscall interface, which looks similar to posix and has functions with similar names, but their interface is fundamentally different, as it wraps over the kernel syscalls.

I apologize for the inconvenience, but if you want to follow a book properly I'm afraid you'd have to either do it in C, or write the necessary libc bindings yourself. I don't think posix interfaces will be in Odin's standard library, at least any time soon. Hopefully the work we're doing on the `os2` package will be good enough to allow to experiment with these concepts, but it won't be a 1:1 correspondence.

**Reply by AxelArielSaravia** [reply]
Thank you so much for your answer.

Yeah, I was aware that the book was fundamentally on POSIX. But I was impressed that Odin has Linux syscalls functions like `msgget()` (and `msg...()`) and `semget()` (and `sem...()`). I haven't seen this in other languages' standard libraries like Go or Zig.

It seems like an issue, but I will explore it more before deciding to send a report because I am not very familiar with Linux/POSIX. Maybe I will create some System V IPC bindings.
<br>
  
> Hopefully the work we're doing on the os2 package will be good enough to allow to experiment with these concepts

This sounds exciting. I will definitely get my hands on it.

**Reply by flysand7** [reply]
> But I was impressed that Odin has Linux syscalls functions like msgget()

I was bored so I just made bindings for all of them
